hledger-lib/Hledger/Data/Amount.hs

```diff
@@ -702,34 +710,32 @@ maIsNonZero = not . mixedAmountIsZero
 --
 -- * the special "missing" mixed amount remains unchanged
 --
-normaliseMixedAmount :: MixedAmount -> MixedAmount
-normaliseMixedAmount = normaliseHelper False
-
-normaliseHelper :: Bool -> MixedAmount -> MixedAmount
-normaliseHelper squashprices (Mixed as)
-  | missingkey `M.member` amtMap = missingmixedamt -- missingamt should always be alone, but detect it even if not
-  | M.null nonzeros = Mixed [newzero]
-  | otherwise       = Mixed $ toList nonzeros
+amounts :: MixedAmount -> [Amount]
+amounts (Mixed ma)
+  | missingkey `M.member` ma = [missingamt]  -- missingamt should always be alone, but detect it even if not
+  | M.null nonzeros          = [newzero]
+  | otherwise                = toList nonzeros
   where
     newzero = fromMaybe nullamt $ find (not . T.null . acommodity) zeros
-    (zeros, nonzeros) = M.partition amountIsZero amtMap
-    amtMap = foldr (\a -> M.insertWith sumSimilarAmountsUsingFirstPrice (key a) a) mempty as
-    key Amount{acommodity=c,aprice=p} = (c, if squashprices then Nothing else priceKey <$> p)
-      where
-        priceKey (UnitPrice  x) = (acommodity x, Just $ aquantity x)
-        priceKey (TotalPrice x) = (acommodity x, Nothing)
-    missingkey = key missingamt
+    (zeros, nonzeros) = M.partition amountIsZero ma
+    missingkey = amountKey missingamt
+
+-- | Get a mixed amount's component amounts without normalising zero and missing
+-- amounts.
+amountsRaw :: MixedAmount -> [Amount]
+amountsRaw (Mixed ma) = toList ma
```

Comment:
I guess we need to export this for JSON and balancing.. is toList's ordering reliable ?

Reply:
It is reliable if we use a `Map` rather than a `HashMap`: it will have the same ordering as the keys.

Reply:
I imagine someone refactoring in future could easily overlook ordering here and start generating different/variable JSON output as a result. I feel we should say something about it in the haddock, specify some kind of invariant. The order of amounts will be... what ? Determined by the ordering of MixedAmountKey, I guess. Which means.. amounts will be ordered by commodity symbol (alphabetically), then by price type (none, unit, total), then by price amount (from most negative to most positive) ?

Reply:
PS and I probably got that ordering wrong - I have now checked MixedAmountKey and I think we could still clarify that MixedAmountKey haddock a bit to make the ordering more obvious.

Reply:
Sure, I'll add some more comments to both of those.